restore-hooks_product-requirements.md

```diff
@@ -0,0 +1,376 @@
+
+**<span style="text-decoration:underline;">USE CASE 3</span>**
+
+
+    **Title: **Run restore at pod level.
+
+
+    **Description: **As a user, I would like to make sure that I can run the restore hook at the pod level. And, I would like the option to run this with an annotation flag in line or using an init container.**<span style="text-decoration:underline;"> </span>**The restore pre-hook should allow the user to run the command on the container where the pre-hook should be executed. Similar to the backup hooks, this hook should run to default to fun on the first container in the pod. 
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 4</span>**
+
+
+    **Title:** Specify container in which to run pre-hook
+
+
+    **Description: **As a user, if I do not want to run the pre-hook command on the first container in the pod (default container), I would like the option to annotate the specific container that i would like the hook to run in.
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 5</span>**
+
+
+    **Title: **Check for latest snapshot 
+
+
+    **Description: **As a user, I would like Velero to run a check for the latest snapshot in object storage prior to starting restore operations on a pod. 
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 6</span>**
+
+
+    **Title: **Specify backup snapshot to use for restore
+
+
+    **Description: **As a user, I would like the option to specify a specific backup snapshot for use by Velero during restore create. I would like to do this instead of using the default most recent backup restore snapshot.
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 7</span>**
+
+
+    **Title: **Include or exclude namespaces from restore
+
+
+    **Description: **As a user, I would like to include all resources in namespaces contained in a backup  by default in my restore spec. 
+
+
+    If velero is asked to restore something that already exists in a pod, the restore will not return a success but will still work - Ashish needs to verify.
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 8</span>**
+
+
+    **Title: **Restore of a stateful application (Unquiescing data from a quiesced backup)
+
+
+    **Description: **As a user, I would like to prepare a backup to be restored without having to freeze an in-use volume/volume that an application is using.
```

Comment:
The first sentence seems tied to backup hooks; is that considered in scope for this work in v1.5?

Reply:
@nrb the scope of this document is for restore only and not backup. However, restore operations rely on some completed aspects of back up. Am I correct in assuming that Velero already quiesces data in a backup? And if that is the case, then can the restore operation here unquiesce the data from a quiesced backup? If that is not the case, then data unquiescing during restore operations may carry a dependency that needs to be documented, and tied to and worked on for backup operations so that we can achieve this use case as part of a restore.

Reply:
>Am I correct in assuming that Velero already quiesces data in a backup?

Velero relies on backup hooks to quiesce application data if it's more than "simple" volume data.

Ideally, applications would include some form of utility that can quiesce data within them that doesn't require freezing or locking.  As an example, `pg_dump` for the PostgreSQL database is an included utility that queiseces the data from the running database without interrupting read/write operations of the live system. Thus, a user could write a hook relying on `pg_dump` as a backup hook, and use the resulting output with `pg_restore` in a restore hook.

If such a utility isn't available, applications would be put into a 'read only' mode during the backup, then allowing writes as soon as the backup is done. Both of these actions are something that the backup hooks would be doing. It would mean interruptions to writes, but that time period should (hopefully) be short. Still, the application would be usable in a read mode.


In the worst case, I think Velero often runs into the stated probelm with the restic support, though, due to restic requiring point-in-time snapshots of the file system that don't change while restic is running. We recommend using fsfreeze out of caution, since the data will be consistent when this is done. However, I absolutely agree that the above "read-only" mode would be preferable in a restic scenario as well as volume snapshots.

Some combination of volume snapshots and restic could provide better support for this, but that is a much larger task that is definitely out of scope. As an example, the idea might be to take a volume snaspshot of a volume, mount a copy of that, then run restic on the volume from that copy. But that's also not how we use restic today - restic's an alternative to volume snapshot plugins, and doesn't work in conjunction with them.

I think maybe the use case is accurate, but it is tough for Velero itself to hit without participation from the applications and hooks themselves.

Reply:
From the discussion during the community call on 7/21, this use-case is referring to a live update of data used by an application by restoring the data from a previous backup. This, we agreed, is out of scope for the restore hooks per-se.

Reply:
That's correct Ashish. Discussed further w/ Nolan and Yaron and we're excluding this requirement from the scope.